typing/typemod.ml

```diff
@@ -1842,6 +1874,60 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
   | Pmod_extension ext ->
       raise (Error_forward (Builtin_attributes.error_of_extension ext))
 
+and type_open_decl ?used_slot ?toplevel names env sod =
+  Builtin_attributes.warning_scope sod.popen_attributes
+    (fun () ->
+       type_open_decl_aux ?used_slot ?toplevel names env sod
+    )
+
+and type_open_decl_aux ?used_slot ?toplevel names env od =
+  let loc = od.popen_loc in
+  match od.popen_expr.pmod_desc with
+  | Pmod_ident lid ->
+    let path, newenv =
+      type_open_ ?used_slot ?toplevel od.popen_override env loc lid
+    in
+    let md = { mod_desc = Tmod_ident (path, lid);
+               mod_type = Mty_alias(Mta_absent, path);
+               mod_env = env;
+               mod_attributes = od.popen_expr.pmod_attributes;
+               mod_loc = od.popen_expr.pmod_loc }
+    in
+    let open_descr = {
+      open_expr = md;
+      open_type = [];
+      open_override = od.popen_override;
+      open_env = newenv;
+      open_loc = loc;
+      open_attributes = od.popen_attributes
+    } in
+    open_descr, [], newenv
+  | _ ->
+    let md = type_module true false None env od.popen_expr in
```

Comment:
The `false` here is a soundness bug. If the `open` appears in a functor body then this must be set to `true`. So `type_open_decl` needs to take an `is_functor` parameter. The following example segfaults with the current code, and would make a good test case:

```ocaml
module type T = sig
  type t
  val x : t
  val f : t -> unit
end

module Int = struct
  type t = int
  let x = 42
  let f = print_int
end

module String = struct
  type t = string
  let x = "Forty Two"
  let f = print_endline
end

let r = ref (module Int : T)

module F (X : sig end) = struct
  open struct
    include (val !r)
  end
  type s = t
  let x : s = x
  let f : s -> unit = f
end

module M = struct end

module N = F(M)

let () =
  r := (module String : T)

module O = F(M)

let () =
  O.f N.x
```